source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2058 @@
+            int error = aws_socket_init(new_socket, allocator, &options);
+            if (error) {
+                aws_mem_release(allocator, new_socket);
+                nw_release(task_args->new_connection);
+                if (listener->accept_result_fn) {
+                    listener->accept_result_fn(listener, task_args->error_code, NULL, task_args->user_data);
+                }
+            } else {
+                new_socket->io_handle.data.handle = task_args->new_connection;
+
+                new_socket->io_handle.set_queue = s_client_set_dispatch_queue;
+
+                nw_endpoint_t endpoint = nw_connection_copy_endpoint(task_args->new_connection);
+                const char *hostname = nw_endpoint_get_hostname(endpoint);
+                uint16_t port = nw_endpoint_get_port(endpoint);
+
+                if (hostname != NULL) {
+                    size_t hostname_len = strlen(hostname);
+                    size_t buffer_size = AWS_ARRAY_SIZE(new_socket->remote_endpoint.address);
+                    size_t to_copy = aws_min_size(hostname_len, buffer_size);
+                    memcpy(new_socket->remote_endpoint.address, hostname, to_copy);
+                    new_socket->remote_endpoint.port = port;
+                }
+                nw_release(endpoint);
+
+                struct nw_socket *new_nw_socket = new_socket->impl;
+                new_nw_socket->os_handle.nw_connection = task_args->new_connection;
+                new_nw_socket->connection_setup = true;
+
+                // Setup socket state to start read/write operations.
+                s_set_socket_state(new_nw_socket, new_socket, CONNECTED_READ | CONNECTED_WRITE);
+
+                nw_connection_set_state_changed_handler(
+                    new_socket->io_handle.data.handle, ^(nw_connection_state_t state, nw_error_t error) {
+                      s_schedule_connection_state_changed_fn(
+                          new_socket, new_nw_socket, new_nw_socket->os_handle.nw_connection, state, error);
+                    });
+
+                // released when the connection state changed to nw_connection_state_cancelled
+                s_socket_acquire_internal_ref(new_nw_socket);
+
+                AWS_LOGF_TRACE(
+                    AWS_LS_IO_SOCKET,
+                    "id=%p handle=%p: incoming connection connected to %s:%d, the incoming handle is %p",
+                    (void *)listener,
+                    listener->io_handle.data.handle,
+                    new_socket->remote_endpoint.address,
+                    new_socket->remote_endpoint.port,
+                    new_socket->io_handle.data.handle);
+
+                if (listener->accept_result_fn) {
+                    listener->accept_result_fn(listener, task_args->error_code, new_socket, task_args->user_data);
+                } else // The connection is not sent to user, clean it up. The nw_connection should be released in
```

Comment:
How would this happen?  Why would we do all this work and then give up at the last instant because of something we could have checked at the beginning?